File: cpp/n257_TreePath.cpp
// https://leetcode-cn.com/problems/binary-tree-paths/
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
   vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        string str="";
        dfs(root,ans,str);
        return ans;
    }
    void dfs(TreeNode *p,vector<string>&ans,string &s) 
    {
        int len = s.length();
        if(p!=NULL)
        {
            if(len == 0)
            s += "";////保证开头没有->
            else
            s += "->";
            s += to_string(p->val);
        }
        else return;
        if(p->left == NULL && p->right == NULL)
        {
            ans.push_back(s);
        }
        else
        dfs(p->left,ans,s),dfs(p->right,ans,s);
        s.erase(s.begin() + len,s.end());//回溯，删除增加的字符，包括->
       
    }
};

int main(){
    return 0;
}

File: python/n257_TreePath.py
# https://leetcode-cn.com/problems/binary-tree-paths/
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode):
        path = []
        cur_path = ""
        def trans_path(root, cur_path, path):
            if root is None: return
            cur_path = cur_path + "->" + str(root.val)
            if root.left is None and root.right is None:
                path.append(cur_path[2:])
                cur_path = cur_path[:-4]
                return
            trans_path(root.left, cur_path, path)
            trans_path(root.right, cur_path, path)
        trans_path(root, cur_path, path)
        return path 


if __name__ == "__main__":
    pass
            